### PR TITLE
Fix errors during the upgrade or downgrade of symfony/flex

### PR DESCRIPTION
**Pull request by rosier**
Composer fails during the upgrade of symfony/flex
```
$ composer create-project symfony/skeleton flex && cd flex
$ composer require symfony/flex dev-main -v
./composer.json has been updated
Running composer update symfony/flex
> pre-update-cmd: Symfony\Flex\Flex_composer_tmp0->configureInstaller
Loading composer repositories with package information
> pre-pool-create: Symfony\Flex\Flex_composer_tmp0->truncatePackages
Updating dependencies
Dependency resolution completed in 0.000 seconds
Analyzed 135 packages to resolve dependencies
Analyzed 210 rules to resolve dependencies
Lock file operations: 0 installs, 1 update, 0 removals
Updates: symfony/flex:dev-main 4d7970f
  - Upgrading symfony/flex (v1.12.2 => dev-main 4d7970f)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
Updates: symfony/flex:dev-main 4d7970f
  - Upgrading symfony/flex (v1.12.2 => dev-main 4d7970f): Extracting archive
PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Symfony\Flex\Command\GenerateIdCommand::__construct(), 0 passed in phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(213) : eval()'d code on line 278 and exactly 1 expected in /private/tmp/flex/vendor/symfony/flex/src/Command/GenerateIdCommand.php:22
Stack trace:
#0 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(213) : eval()'d code(278): Symfony\Flex\Command\GenerateIdCommand->__construct()
#1 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(320): Symfony\Flex\Flex_composer_tmp1->activate(Object(Composer\Composer), Object(Composer\IO\ConsoleIO))
#2 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(224): Composer\Plugin\PluginManager->addPlugin(Object(Symfony\Flex\Flex_composer_tmp1), false)
#3 phar:///usr/local/bin/composer/src/Composer/Installer/PluginInstaller.php(100): Composer\Plugin\PluginManager->registerPackage(Object(Composer\Package\CompletePackage), true)
#4 [in in /private/tmp/flex/vendor/symfony/flex/src/Command/GenerateIdCommand.php on line 22

### CODE REVIEW

**Comment by rosier** [comment]
Composer also fails during the downgrade of symfony/flex
```
$ composer create-project symfony/skeleton flex && cd flex
$ composer require symfony/flex dev-main --no-plugins --no-scripts
$ composer require symfony/flex
Using version ^1.12 for symfony/flex
./composer.json has been updated
Running composer update symfony/flex
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 1 update, 0 removals
  - Upgrading symfony/flex (dev-main 4d7970f => v1.12.2)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Upgrading symfony/flex (dev-main 4d7970f => v1.12.2): Extracting archive
PHP Fatal error:  Uncaught Error: Call to undefined method Symfony\Flex\Downloader::setFlexId() in phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(213) : eval()'d code:129
Stack trace:
#0 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(320): Symfony\Flex\Flex_composer_tmp1->activate(Object(Composer\Composer), Object(Composer\IO\ConsoleIO))
#1 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(224): Composer\Plugin\PluginManager->addPlugin(Object(Symfony\Flex\Flex_composer_tmp1), false)
#2 phar:///usr/local/bin/composer/src/Composer/Installer/PluginInstaller.php(100): Composer\Plugin\PluginManager->registerPackage(Object(Composer\Package\CompletePackage), true)
#3 [internal function]: Composer\Installer\PluginInstaller->Composer\Installer\{closure}(NULL)
#4 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/FulfilledPromise.php(20): call_user_func(Object(Closure), NULL)
#5 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred in phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(213) : eval()'d code on line 129

File: src/Command/GenerateIdCommand.php [hunk 1]
```diff
@@ -18,6 +18,13 @@
 
 class GenerateIdCommand extends Command
 {
+    public function __construct(/* cannot be type-hinted */ $flex)
```
Comment (stof) [inline src/Command/GenerateIdCommand.php:21]: the argument should be optional

Comment (rosier) [reply src/Command/GenerateIdCommand.php]: done

File: src/Flex.php [hunk 2]
```diff
@@ -275,7 +275,7 @@ class_exists(__NAMESPACE__.str_replace('/', '\\', substr($file, \strlen(__DIR__)
             $app->add(new Command\UnpackCommand($resolver));
             $app->add(new Command\RecipesCommand($this, $this->lock, $rfs));
             $app->add(new Command\InstallRecipesCommand($this, $this->options->get('root-dir')));
-            $app->add(new Command\GenerateIdCommand());
+            $app->add(new Command\GenerateIdCommand(null));
```
Comment (fabpot) [inline src/Flex.php:278]: To prepare the next step, we can maybe register the command only if it exists?

```php
if (class_exists(Command\GenerateIdCommand::class)) {
    $app->add(new Command\GenerateIdCommand(null));
}
```

Comment (rosier) [reply src/Flex.php:278]: Indeed, that should clear the way to remove the command later

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @rosier.

**Pull request merged**